include/unifex/task.hpp

```diff
@@ -273,29 +194,58 @@ struct _task<T>::type {
   }
 
 private:
+  template<typename Promise>
   struct awaiter {
-    coro::coroutine_handle<promise_type> coro_;
+    explicit awaiter(coro::coroutine_handle<promise_type> coro) noexcept
+    : coro_(coro)
+    {}
+
+    awaiter(awaiter&& other) noexcept
+    : coro_(std::exchange(other.coro_, {}))
+    {}
+
+    ~awaiter() {
+      if (coro_) coro_.destroy();
+    }
+
     bool await_ready() noexcept {
       return false;
     }
-    template <typename OtherPromise>
-    auto await_suspend(
-        coro::coroutine_handle<OtherPromise> h) noexcept {
+
+    coro::coroutine_handle<promise_type> await_suspend(
+        coro::coroutine_handle<Promise> h) noexcept {
+      assert(coro_);
       coro_.promise().continuation_ = h;
       coro_.promise().info_.emplace(
           continuation_info::from_continuation(h.promise()));
+      coro_.promise().stoken_ = stopTokenAdapter_.subscribe(get_stop_token(h.promise()));
       return coro_;
     }
-    decltype(auto) await_resume() {
+
+    std::optional<non_void_t<wrap_reference_t<decay_rvalue_t<T>>>> await_resume() {
```

Comment:
This is a surprising choice, I think.  It makes an uncancellable task inexpressible, right?

Reply:
An uncancellable task will just never complete with `nullopt`.

I was going for a consistent shape to the result of `co_await` expressions - ie. every `co_await` in a coroutine completes in a `std::optional`.
This allows the `connect_awaitable()` implementation to always be able to unpack the result of the `co_await` and map an empty `optional` to `set_done()`.